### ISSUE TITLE
SerializationHelper.SerializeFormModel doesn't seem to work properly

### ISSUE THREAD

**Issue by johnmasson** [kjac/FormEditor#189]
Serializing the form model with TypeNameHandling.Auto serializes the Pages property into a object like this (after processing):

```
"pages":{  
      "runtimeType":"FormEditor.Page[], FormEditor",
      "$values":[  ]
   }
```
which doesn't load into the editor correctly. 

I needed to use TypeNameHandling.Objects to get JSON which would both load into the editor and deserialize correctly into a valid FormModel using strongly typed content properties. This results in a lot of unecessary '$type' properties on the JSON objects but does seem to work properly.

I'm using the latest NuGet package with Umbraco 7.8.1.

**Comment by johnmasson** [comment]
I'm creating forms from code, as per your example:

https://github.com/kjac/FormEditor/blob/master/Docs/initialize.md

The bug is that SerializationHelper creates JSON which then doesn't load into the editor (if used as in the example).

**Comment by johnmasson** [comment]
Reading this back, I realize I might be being unclear. Steps to reproduce the problem:

1. Create a form programmatically, as per the example in the docs.
2. Access the resulting content node in the Umbraco backend. The form appears empty, with none of the rows, fields etc. added from code showing up in the UI. Inspecting the form editors $scope.model.value shows the pages property as described above, with the form content in the $values array where the UI can't find it.

I am not attempting to deserialize the form property from my own code, or in any other way interfering with Umbraco's normal operations at that point.

**Comment by kjac** [comment]
@johnmasson aah... makes sense. Are you using the `SerializationHelper` to serialize the `FormModel` when creating your content, as per [this tutorial](https://github.com/kjac/FormEditor/blob/0adf5a1c80d476ff51207de09f58b2e0607a6a39/Docs/initialize.md#creating-the-form-programmatically)?

**Comment by johnmasson** [comment]
Yes, I am.

**Comment by kjac** [comment]
Strange. I'll try to find time to reproduce it this weekend.

**Comment by kjac** [comment]
Hi @johnmasson,

I can't reproduce this. I've used Umbraco 7.9.2 instead of 7.8.1, but that shouldn't cause any difference.

I've used this event handler (copied from the above mentioned tutorial):

```cs
using System.Collections.Generic;
using FormEditor;
using FormEditor.Fields;
using Umbraco.Core;
using Umbraco.Core.Events;
using Umbraco.Core.Models;
using Umbraco.Core.Services;

namespace Umbraco79
{
  public class UmbracoEventsHandler : ApplicationEventHandler
  {
    protected override void ApplicationStarted(UmbracoApplicationBase umbracoApplication, ApplicationContext applicationContext)
    {
      // hook into the Created event
      ContentService.Created += ContentServiceOnCreated;
    }

    private void ContentServiceOnCreated(IContentService sender, NewEventArgs<IContent> newEventArgs)
    {
      // if creating content of type "myFormContentType", add a default form
      if (newEventArgs.Alias == "myFormContentType")
      {
        // create a one page form containing two rows:
        // - one two-column row with a name and an email field
        // - one one-column row with a message field and a submit button
        var form = new FormModel
        {
          Pages = new List<Page>
          {
            new Page
            {
              Rows = new List<Row>
              {
                // the row and cell aliases used below are the default aliases from the 
                // Form Editor data type - replace them with your own
                new Row
                {
                  Alias = "two-column",
                  Cells = new List<Cell>
                  {
                    new Cell
                    {
                      Alias = "col-md-6",
                      Fields = new List<Field>
                      {
                        new TextBoxField
                        {
                          Name = "Name",
                          Label = "Your name",
                          ErrorMessage = "Please enter your name",
                          Placeholder = "Enter your name",
                          Mandatory = true
                        }
                      }
                    },
                    new Cell
                    {
                      Alias = "col-md-6",
                      Fields = new List<Field>
                      {
                        new EmailField
                        {
                          Name = "Email",
                          Label = "Your email",
                          ErrorMessage = "Please enter your email",
                          Placeholder = "Enter your email",
                          Mandatory = false,
                          HelpText = "Your email will not be displayed on the site."
                        }
                      }
                    }

                  }
                },
                new Row
                {
                  Alias = "one-column",
                  Cells = new List<Cell>
                  {
                    new Cell
                    {
                      Alias = "col-md-12",
                      Fields = new List<Field>
                      {
                        new TextAreaField
                        {
                          Name = "Message",
                          Label = "Your message",
                          ErrorMessage = "Please enter your message",
                          Placeholder = "Enter your message",
                          Mandatory = true
                        },
                        new SubmitButtonField
                        {
                          Text = "Send your message"
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        };

        // serialize the form to JSON using the Form Editor serialization helper
        // and store the JSON in the "form" property
        newEventArgs.Entity.SetValue("form", SerializationHelper.SerializeFormModel(form));
      }
    }
  }
}
```

It yields the following JSON:

```json
{
  "pages": [{
      "rows": [{
          "alias": "two-column",
          "cells": [{
              "alias": "col-md-6",
              "fields": [{
                  "runtimeType": "FormEditor.Fields.TextBoxField, FormEditor",
                  "prettyName": "Text box",
                  "type": "core.textbox",
                  "maxLength": 0,
                  "placeholder": "Enter your name",
                  "mandatory": true,
                  "errorMessage": "Please enter your name",
                  "label": "Your name",
                  "name": "Name",
                  "hasSubmittedValue": false,
                  "supportsStripHtml": true,
                  "view": "core.textbox.html",
                  "icon": "core.textbox.png"
                }
              ]
            }, {
              "alias": "col-md-6",
              "fields": [{
                  "runtimeType": "FormEditor.Fields.EmailField, FormEditor",
                  "prettyName": "Email address",
                  "type": "core.email",
                  "multiple": false,
                  "placeholder": "Enter your email",
                  "mandatory": false,
                  "errorMessage": "Please enter your email",
                  "label": "Your email",
                  "helpText": "Your email will not be displayed on the site.",
                  "name": "Email",
                  "hasSubmittedValue": false,
                  "supportsStripHtml": true,
                  "view": "core.email.html",
                  "icon": "core.email.png"
                }
              ]
            }
          ]
        }, {
          "alias": "one-column",
          "cells": [{
              "alias": "col-md-12",
              "fields": [{
                  "runtimeType": "FormEditor.Fields.TextAreaField, FormEditor",
                  "prettyName": "Text area",
                  "type": "core.textarea",
                  "maxLength": 0,
                  "placeholder": "Enter your message",
                  "mandatory": true,
                  "errorMessage": "Please enter your message",
                  "label": "Your message",
                  "name": "Message",
                  "hasSubmittedValue": false,
                  "supportsStripHtml": true,
                  "view": "core.textarea.html",
                  "icon": "core.textarea.png"
                }, {
                  "runtimeType": "FormEditor.Fields.SubmitButtonField, FormEditor",
                  "type": "core.submitbutton",
                  "prettyName": "Submit button",
                  "text": "Send your message",
                  "view": "core.submitbutton.html",
                  "icon": "core.submitbutton.png"
                }
              ]
            }
          ]
        }
      ]
    }
  ],
  "rows": [{
      "alias": "two-column",
      "cells": [{
          "alias": "col-md-6",
          "fields": [{
              "runtimeType": "FormEditor.Fields.TextBoxField, FormEditor",
              "prettyName": "Text box",
              "type": "core.textbox",
              "maxLength": 0,
              "placeholder": "Enter your name",
              "mandatory": true,
              "errorMessage": "Please enter your name",
              "label": "Your name",
              "name": "Name",
              "hasSubmittedValue": false,
              "supportsStripHtml": true,
              "view": "core.textbox.html",
              "icon": "core.textbox.png"
            }
          ]
        }, {
          "alias": "col-md-6",
          "fields": [{
              "runtimeType": "FormEditor.Fields.EmailField, FormEditor",
              "prettyName": "Email address",
              "type": "core.email",
              "multiple": false,
              "placeholder": "Enter your email",
              "mandatory": false,
              "errorMessage": "Please enter your email",
              "label": "Your email",
              "helpText": "Your email will not be displayed on the site.",
              "name": "Email",
              "hasSubmittedValue": false,
              "supportsStripHtml": true,
              "view": "core.email.html",
              "icon": "core.email.png"
            }
          ]
        }
      ]
    }, {
      "alias": "one-column",
      "cells": [{
          "alias": "col-md-12",
          "fields": [{
              "runtimeType": "FormEditor.Fields.TextAreaField, FormEditor",
              "prettyName": "Text area",
              "type": "core.textarea",
              "maxLength": 0,
              "placeholder": "Enter your message",
              "mandatory": true,
              "errorMessage": "Please enter your message",
              "label": "Your message",
              "name": "Message",
              "hasSubmittedValue": false,
              "supportsStripHtml": true,
              "view": "core.textarea.html",
              "icon": "core.textarea.png"
            }, {
              "runtimeType": "FormEditor.Fields.SubmitButtonField, FormEditor",
              "type": "core.submitbutton",
              "prettyName": "Submit button",
              "text": "Send your message",
              "view": "core.submitbutton.html",
              "icon": "core.submitbutton.png"
            }
          ]
        }
      ]
    }
  ],
  "rowId": "00000000-0000-0000-0000-000000000000",
  "emailNotificationAttachments": false,
  "successPageId": 0,
  "disallowMultipleSubmissionsPerUser": false,
  "addSubmissionInfo": false,
  "maxSubmissionsForCurrentUserHandler": {
    "runtimeType": "FormEditor.Limitations.MaxSubmissionsForCurrentUserHandler, FormEditor"
  }
}
```

...and looks like this in the backoffice:

![189](https://user-images.githubusercontent.com/7405322/39661926-81051f16-505a-11e8-84fd-708e021f4313.png)

**Comment by johnmasson** [comment]
OK, I've figured it out: although FormModel.Pages is declared as IEnumerable<Page> and accepts any implementation of IEnumerable<Page> in the constructor, it doesn't serialize properly when the concrete type is an array (as it was in my code) instead of List<Page> (as in your example). It looks as if the same is true for Page.Rows, Rows.Cells etc. I've updated my code to use List<> everywhere and it now works correctly with your SerializationHelper. Thanks for looking into this.
